apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -111,39 +121,55 @@ internal_resolve(Protocol, DNSMessage) ->
     Response = erldns_handler:do_handle(DNSMessage, ?LOCALHOST),
     encode_message(Protocol, Response).
 
--spec(resolve(protocol(), [upstream()], binary(), reply_fun()) -> ok).
-resolve(Protocol, Upstreams, Request, Fun) ->
+-spec(resolve(
+    pos_integer(),
+    protocol(),
+    [upstream()],
+    binary(),
+    reply_fun(),
+    binary(),
+    binary()) -> ok).
+resolve(Begin, Protocol, Upstreams, Request, Fun, FailMsg, Zone) ->
     Workers =
         lists:map(fun (Upstream) ->
             Pid = start_worker(Protocol, Upstream, Request),
             MonRef = monitor(process, Pid),
             {Upstream, Pid, MonRef}
         end, Upstreams),
-    resolve_loop(Workers, Fun).
-
--spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun()) -> ok).
-resolve_loop([], _Fun) ->
+    resolve_loop(Begin, Workers, Fun, FailMsg, Zone).
+
+-spec(resolve_loop(
+    pos_integer(),
+    [{upstream(), pid(), reference()}],
+    reply_fun(),
+    binary(),
+    binary()) -> ok).
+resolve_loop(_Begin, [], skip, _FailMsg, _Zone) ->
     ok;
-resolve_loop(Workers, Fun) ->
+resolve_loop(Begin, [], Fun, FailMsg, Zone) ->
+    prometheus_counter:inc(dns_forwarder_failures_total, [Zone]),
```

Comment:
@urbanserj I am not sure about here. I inverted the lines as when you get to this point, we should emit `failure` regardless of `ok = reply...` working or not.